File: macos/DotsHarness/Sources/HarnessPluginKit/Composition.swift
// Copyright (c) 2026 DOTS
// Standalone plugin composition model for Dots Harness.

import Foundation

public struct CompositionEntry: Codable, Sendable, Equatable {
    public var id: String
    public var plugin: String
    public var disabled: Bool
    public var isolate: [String: Bool]
    public var config: JSONObject

    public init(
        id: String,
        plugin: String,
        disabled: Bool = false,
        isolate: [String: Bool] = [:],
        config: JSONObject = [:]
    ) {
        self.id = id
        self.plugin = plugin
        self.disabled = disabled
        self.isolate = isolate
        self.config = config
    }

    enum CodingKeys: String, CodingKey {
        case id, plugin, disabled, isolate, config
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decode(String.self, forKey: .id)
        plugin = try container.decode(String.self, forKey: .plugin)
        disabled = try container.decodeIfPresent(Bool.self, forKey: .disabled) ?? false
        isolate = try container.decodeIfPresent([String: Bool].self, forKey: .isolate) ?? [:]
        config = try container.decodeIfPresent(JSONObject.self, forKey: .config) ?? [:]
    }
}

public struct CompositionDocument: Codable, Sendable, Equatable {
    public var plane: PluginPlane
    public var entries: [CompositionEntry]

    public init(plane: PluginPlane, entries: [CompositionEntry]) {
        self.plane = plane
        self.entries = entries
    }
}

public enum CompositionPatch: Sendable, Equatable {
    case insert([CompositionEntry])
    case disable(String)
    case enable(String)
    case mergeConfig(id: String, config: JSONObject)

    public static func apply(_ patches: [CompositionPatch], to document: CompositionDocument) -> CompositionDocument {
        var entries = document.entries
        for patch in patches {
            switch patch {
            case .insert(let incoming):
                for entry in incoming {
                    if let index = entries.firstIndex(where: { $0.id == entry.id }) {
                        entries[index] = entry
                    } else {
                        entries.append(entry)
                    }
                }
            case .disable(let id):
                if let index = entries.firstIndex(where: { $0.id == id }) {
                    entries[index].disabled = true
                }
            case .enable(let id):
                if let index = entries.firstIndex(where: { $0.id == id }) {
                    entries[index].disabled = false
                }
            case .mergeConfig(let id, let config):
                if let index = entries.firstIndex(where: { $0.id == id }) {
                    entries[index].config.merge(from: config)
                }
            }
        }
        return CompositionDocument(plane: document.plane, entries: entries)
    }
}

public struct MountIssue: Sendable, Equatable, Identifiable {
    public var id: String { rowId + ":" + message }
    public var rowId: String
    public var message: String

    public init(rowId: String, message: String) {
        self.rowId = rowId
        self.message = message
    }
}

public enum PluginError: Error, Sendable, Equatable, LocalizedError {
    case missingService(String)
    case unknownPlugin(String)
    case incompatibleABI(String)
    case untrustedLibrary(String)
    case applyFailed(String)
    case globalService(String)
    case slotCollision(String)
    case invalidManifest(String)
    case invalidComposition(String)
    case package(String)

    public var errorDescription: String? {
        switch self {
        case .missingService(let name): return "waiting for \(name)"
        case .unknownPlugin(let id): return "Cannot find plugin \(id)"
        case .incompatibleABI(let abi): return "incompatible ABI \(abi)"
        case .untrustedLibrary(let id): return "untrusted dylib \(id)"
        case .applyFailed(let message): return message
        case .globalService(let name): return "row published process-global service [\(name)]; a session service must sit behind an isolate realm or move to the host composition"
        case .slotCollision(let message): return message
        case .invalidManifest(let message): return message
        case .invalidComposition(let message): return message
        case .package(let message): return message
        }
    }
}

File: windows/DotsHarness/src/HarnessPluginKit/Composition.cs
// Copyright (c) 2026 DOTS
// Plugin composition model derived from DeepSeek Harness.
// Copyright (c) 2026 DeepSeek. MIT. See NOTICE.

namespace HarnessPluginKit;

public sealed record CompositionEntry(
    string Id,
    string Plugin,
    bool Disabled = false,
    IReadOnlyDictionary<string, bool>? Isolate = null,
    IReadOnlyDictionary<string, JsonValue>? Config = null)
{
    public IReadOnlyDictionary<string, bool> Isolate { get; init; } =
        Isolate ?? new Dictionary<string, bool>(StringComparer.Ordinal);

    public IReadOnlyDictionary<string, JsonValue> Config { get; init; } =
        Config ?? new Dictionary<string, JsonValue>(StringComparer.Ordinal);

    public CompositionEntry WithDisabled(bool disabled) => this with { Disabled = disabled };

    public CompositionEntry WithMergedConfig(IReadOnlyDictionary<string, JsonValue> incoming) =>
        this with { Config = Config.MergeFrom(incoming) };
}

public sealed record CompositionDocument(PluginPlane Plane, IReadOnlyList<CompositionEntry> Entries)
{
    public CompositionDocument WithEntries(IReadOnlyList<CompositionEntry> entries) => this with { Entries = entries };
}

public abstract record CompositionPatch
{
    public sealed record Insert(IReadOnlyList<CompositionEntry> Entries) : CompositionPatch;
    public sealed record Disable(string Id) : CompositionPatch;
    public sealed record Enable(string Id) : CompositionPatch;
    public sealed record MergeConfig(string Id, IReadOnlyDictionary<string, JsonValue> Config) : CompositionPatch;

    public static CompositionDocument Apply(IReadOnlyList<CompositionPatch> patches, CompositionDocument document)
    {
        var entries = document.Entries.ToList();
        foreach (var patch in patches)
        {
            switch (patch)
            {
                case Insert insert:
                    foreach (var entry in insert.Entries)
                    {
                        var index = entries.FindIndex(e => e.Id == entry.Id);
                        if (index >= 0) entries[index] = entry;
                        else entries.Add(entry);
                    }
                    break;
                case Disable disable:
                    Replace(entries, disable.Id, e => e.WithDisabled(true));
                    break;
                case Enable enable:
                    Replace(entries, enable.Id, e => e.WithDisabled(false));
                    break;
                case MergeConfig merge:
                    Replace(entries, merge.Id, e => e.WithMergedConfig(merge.Config));
                    break;
            }
        }
        return document.WithEntries(entries);
    }

    private static void Replace(List<CompositionEntry> entries, string id, Func<CompositionEntry, CompositionEntry> map)
    {
        var index = entries.FindIndex(e => e.Id == id);
        if (index >= 0) entries[index] = map(entries[index]);
    }
}

public sealed record MountIssue(string RowId, string Message)
{
    public string Id => RowId + ":" + Message;
}

public sealed class PluginException : Exception
{
    public PluginErrorKind Kind { get; }

    public PluginException(PluginErrorKind kind, string message) : base(message)
    {
        Kind = kind;
    }

    public static PluginException MissingService(string name) =>
        new(PluginErrorKind.MissingService, $"waiting for {name}");

    public static PluginException UnknownPlugin(string id) =>
        new(PluginErrorKind.UnknownPlugin, $"Cannot find plugin {id}");

    public static PluginException IncompatibleAbi(string abi) =>
        new(PluginErrorKind.IncompatibleAbi, $"incompatible ABI {abi}");

    public static PluginException UntrustedLibrary(string id) =>
        new(PluginErrorKind.UntrustedLibrary, $"untrusted dylib {id}");

    public static PluginException ApplyFailed(string message) =>
        new(PluginErrorKind.ApplyFailed, message);

    public static PluginException GlobalService(string name) =>
        new(PluginErrorKind.GlobalService,
            $"row published process-global service [{name}]; a session service must sit behind an isolate realm or move to the host composition");

    public static PluginException SlotCollision(string message) =>
        new(PluginErrorKind.SlotCollision, message);

    public static PluginException InvalidManifest(string message) =>
        new(PluginErrorKind.InvalidManifest, message);

    public static PluginException InvalidComposition(string message) =>
        new(PluginErrorKind.InvalidComposition, message);
}

public enum PluginErrorKind
{
    MissingService,
    UnknownPlugin,
    IncompatibleAbi,
    UntrustedLibrary,
    ApplyFailed,
    GlobalService,
    SlotCollision,
    InvalidManifest,
    InvalidComposition,
}
